atom/browser/api/atom_api_app.cc

```diff
@@ -262,6 +263,36 @@ v8::Local<v8::Value> App::DefaultSession(v8::Isolate* isolate) {
     return v8::Local<v8::Value>::New(isolate, default_session_);
 }
 
+bool App::OnProcessSingletonNotification(
+    const base::CommandLine& command_line,
+    const base::FilePath& current_directory) {
+  ProcessSingleton::NotificationCallback cb;
+  mate::Converter<ProcessSingleton::NotificationCallback>::FromV8(isolate(), single_instance_callback_, &cb);
+  
+  return cb.Run(command_line, current_directory);
+}
+
+bool App::MakeSingleInstance(v8::Local<v8::Function> callback) {
```

Comment:
I just checked the source code of `ProcessSingleton` and it is ensuring the callback is called on UI thread, and you also have to create it on UI thread too.